### ISSUE TITLE
Disallow defining the same field twice in a single mapping definition.

### ISSUE THREAD

**Issue by jtibshirani** [elastic/elasticsearch#37437]
In #37141 we discovered that it is possible to define the same field twice in a single request to add mappings, so long as the field mapping definitions are compatible and can be merged together. For example, the following mapping is valid, and the first definition happens to 'win out':

```
PUT index
{
  "mappings": {
    "_doc": {
      "properties": {
          "nested": {
              "properties": {
                  "field": {
                      "type":"keyword"
                }
              }
          },
          "nested.field": {
              "type":"keyword",
              "boost": 3.0
          }
      }
    }
  }
}
```

To me, a request like this is not likely to be intentional, and could indicate a bug in the calling code. **For this reason it may make sense to detect this case and disallow it.**

As a note, the above mapping update would be rejected if the duplicate field were specified at the same level. For example, the following mapping fails with a JSON parsing exception:

```
PUT index
{
  "mappings": {
    "_doc": {
      "properties": {
          "nested.field": {
              "type":"keyword"
          },
          "nested.field": {
              "type":"keyword"
          }
      }
    }
  }
}
```

**Comment by zacharymorn** [comment]
This issue has been open for more than a year. Does it still need fixing?

**Comment by jtibshirani** [comment]
@zacharymorn yes, I still think it'd be valuable to fix. Let me know if you're interested in looking into it. As a heads up, there are some details that still need settling like whether we need a deprecation period where we warn about the duplicate entries before we outright disallow them.

**Comment by zacharymorn** [comment]
@jtibshirani I've created a draft PR for this. For the time being I chose to emit deprecation warning when the issue in question is detected. Please let me know if this looks good to you.

**Comment by jpountz** [comment]
I thought we were already checking this here: https://github.com/elastic/elasticsearch/blob/c94ebefbc67f342157ce9e7a8dd7a7fd9d07b409/server/src/main/java/org/elasticsearch/index/mapper/MapperMergeValidator.java#L61. But obviously we're not, I wonder whether this validation logic is buggy.

**Comment by jtibshirani** [comment]
I looked into this a bit more -- the issue is that when building the root `ObjectMapper`, we attempt to merge together fields with the same name: https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java#L143

So if there is no field mapper conflict as in the original example, the merge succeeds and the object mapper will contain a single definition for the field. By the time we call into `MapperMergeValidator.validateNewMappers`, there is only one field mapper.

Although we've seen it come up once before (#51286), to me this feels like confusing enough behavior, and also uncommon enough that we could treat it as a bug instead of a behavior change. Another reason I'm hesitant to treat it as a normal behavior change is that in addition to a deprecation period, we'd want to consider API versioning, which seems too heavy-handed.

Maybe a compromise would be to only make the fix in 8.0.0, and not backport it? For major upgrades, many users already give a close look at their mappings and are ready to make changes.

**Comment by jtibshirani** [comment]
I did some testing and found that if we changed this behavior, we could keep these related behaviors the same:
* A dynamic mapping update would still work given a document like above that defines the same field twice. This is because we merge dynamic mapping updates together in `DocumentParser#createDynamicUpdate` before constructing the combined update.
* As expected, you could still define a field using structured JSON notation, and then later send a mapping update using dot notation, and it would succeed.

A note that at a high level, we rely on this merging behavior for a mapping like the following to work:
```
PUT index
{
  "mappings": {
    "properties": {
      "nested.field": {
        "type": "keyword"
      },
      "nested.other_field": {
        "type": "keyword"
      }
    }
  }
}
```

The change should make sure to preserve the ability to define mappings like this. For this reason, I think the fix will be a bit more complex.

**Comment by jpountz** [comment]
Thanks for digging. I agree we should preserve the ability to define mappings like this.

**Comment by zacharymorn** [comment]
I'm trying to follow the discussion but feel I might have some context/knowledge gap. 

From my quick reading of dynamic mapping, it seems to be trigged by index document update. But I was previously under the impression that the fix to this issue is to detect the nested field definition duplication for explicit index creation or mapping updating. So they seems to be on different code patch.

Am I missing something? Is my PR totally on the wrong direction :D ?

**Comment by jtibshirani** [comment]
@zacharymorn sorry for the late response. Your overall understanding is correct, I'll just add some clarifications. This issue is indeed focused on duplicate field definitions in index creation/ mapping updates. As you say, dynamic mappings are triggered by document updates. Dynamic mappings do have their own code path, but they share quite a bit of logic with normal mapping updates. So @jpountz's question was whether changing this behavior for normal mapping updates could also affect dynamic updates. 

> wouldn’t that create inconsistencies in field definitions from different methods?

This is a good question, and one I had after reading the dynamic update example. I think it could also use discussion internally (in addition to the deprecation question). I'll give some preliminary comments on your PR as we work to sort out these open questions.

**Comment by jtibshirani** [comment]
We met to discuss and have decided the following:
* We indeed want to reject mappings with duplicate keys. For now, we will still accept duplicate keys in documents, although this is inconsistent in a sense.
* We'll issue a deprecation warning about the issue in 7.x, and then in 8.0 we'll reject new mappings with duplicate keys.

Some implementation notes:
* Typically we always develop against master. So we can first create a PR against master that issues deprecation warnings, then backport it to 7.x. After, we'll follow-up with a change on master (8.0) to remove the deprecation and start throwing an exception.
* The change should target the mappings validation/ building logic, not JSON parsing itself.
* We should make sure to preserve the behaviors described in https://github.com/elastic/elasticsearch/issues/37437#issuecomment-610066534.

Lastly for the team's context, here are the two issues I've seen where users ran into this confusing behavior: https://github.com/elastic/elasticsearch/issues/51286, https://github.com/elastic/elasticsearch/issues/37141.

**Comment by Razi007** [comment]
@jtibshirani is this issue is still open?

**Comment by jtibshirani** [comment]
@Razi007 there is already an open PR from @zacharymorn (#54481), I just totally dropped the ball on reviewing it. Since there's already pre-existing work, it'd be best to check out some other issues with the `help wanted` or `good first issue` label.
